File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueComplexidadeLinear.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Berlekamp-Massey: calcula a complexidade linear (tamanho do menor LFSR que
 * reproduz a sequência de bits). Para uma sequência aleatória de N bits, a
 * complexidade fica perto de N/2. Se a cifra esconder estrutura linear (tipo
 * um LFSR disfarçado), a complexidade cai MUITO abaixo disso - e aí a cifra
 * seria atacável resolvendo um sistema linear em vez de força bruta.
 */
public class AtaqueComplexidadeLinear implements AtaqueInterface {
    private final int bitsPorAmostra;
    private final int amostras;
    private final double limiteRelativo;

    public AtaqueComplexidadeLinear() {
        this(1024, 5, 0.4);
    }

    public AtaqueComplexidadeLinear(int bitsPorAmostra, int amostras, double limiteRelativo) {
        this.bitsPorAmostra = bitsPorAmostra;
        this.amostras = amostras;
        this.limiteRelativo = limiteRelativo;
    }

    @Override
    public String nome() {
        return "Complexidade linear (Berlekamp-Massey)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] primeiro = alvo.gerarKeystreamBruto(chave, Util.randomBytes(alvo.tamanhoIv()), "enc", 64);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        double[] relativos = new double[this.amostras];
        double pior = 1.0;
        for (int a = 0; a < this.amostras; a++) {
            byte[] bytes = alvo.gerarKeystreamBruto(
                    chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.bitsPorAmostra / 8);
            int[] bits = this.paraBits(bytes, this.bitsPorAmostra);
            double relativo = (double) this.berlekampMassey(bits) / this.bitsPorAmostra;
            relativos[a] = relativo;
            pior = Math.min(pior, relativo);
        }

        double soma = 0.0;
        for (double r : relativos) {
            soma += r;
        }
        double media = soma / relativos.length;
        boolean vulneravel = pior < this.limiteRelativo;

        Map<String, Object> dados = new HashMap<>();
        dados.put("media", media);
        dados.put("pior", pior);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.CRITICA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "complexidade linear relativa: média=%.1f%%, pior=%.1f%% de %d bits (esperado ~50%%; limite>=%.0f%%)",
                        media * 100,
                        pior * 100,
                        this.bitsPorAmostra,
                        this.limiteRelativo * 100),
                dados);
    }

    private int[] paraBits(byte[] bytes, int limite) {
        int[] bits = new int[limite];
        int idx = 0;
        for (int i = 0; i < bytes.length && idx < limite; i++) {
            int b = bytes[i] & 0xFF;
            for (int j = 7; j >= 0 && idx < limite; j--) {
                bits[idx++] = (b >> j) & 1;
            }
        }
        return bits;
    }

    /** Algoritmo de Berlekamp-Massey sobre GF(2). */
    private int berlekampMassey(int[] s) {
        int n = s.length;
        int[] c = new int[n];
        c[0] = 1;
        int[] b = new int[n];
        b[0] = 1;
        int l = 0;
        int m = -1;

        for (int i = 0; i < n; i++) {
            int d = s[i];
            for (int j = 1; j <= l; j++) {
                d ^= c[j] & s[i - j];
            }

            if (d == 1) {
                int[] t = c.clone();
                int shift = i - m;
                for (int j = 0; j + shift < n; j++) {
                    c[j + shift] ^= b[j];
                }
                if (l <= i / 2) {
                    l = i + 1 - l;
                    m = i;
                    System.arraycopy(t, 0, b, 0, n);
                }
            }
        }

        return l;
    }
}

File: packages/go/internal/seguranca/ataques/ataque_complexidade_linear.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesComplexidadeLinear configura o AtaqueComplexidadeLinear.
type OpcoesComplexidadeLinear struct {
	BitsPorAmostra int
	Amostras       int
	LimiteRelativo float64
}

// AtaqueComplexidadeLinear usa Berlekamp-Massey para calcular a complexidade
// linear (tamanho do menor LFSR que reproduz a sequência de bits).
type AtaqueComplexidadeLinear struct {
	bitsPorAmostra int
	amostras       int
	limiteRelativo float64
}

// NovoAtaqueComplexidadeLinear cria o ataque. Sem argumento, usa
// bitsPorAmostra = 1024, amostras = 5 e limiteRelativo = 0.4.
func NovoAtaqueComplexidadeLinear(opcoes ...OpcoesComplexidadeLinear) *AtaqueComplexidadeLinear {
	o := OpcoesComplexidadeLinear{BitsPorAmostra: 1024, Amostras: 5, LimiteRelativo: 0.4}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueComplexidadeLinear{
		bitsPorAmostra: o.BitsPorAmostra,
		amostras:       o.Amostras,
		limiteRelativo: o.LimiteRelativo,
	}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueComplexidadeLinear) Nome() string {
	return "Complexidade linear (Berlekamp-Massey)"
}

// Executar roda o ataque.
func (a *AtaqueComplexidadeLinear) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())

	relativos := make([]float64, 0, a.amostras)
	pior := 1.0
	for i := 0; i < a.amostras; i++ {
		bytes := alvo.GerarKeystreamBruto(chave, seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.bitsPorAmostra/8)
		bits := a.paraBits(bytes, a.bitsPorAmostra)
		relativo := float64(a.berlekampMassey(bits)) / float64(a.bitsPorAmostra)
		relativos = append(relativos, relativo)
		if relativo < pior {
			pior = relativo
		}
	}

	media := 0.0
	for _, v := range relativos {
		media += v
	}
	media /= float64(len(relativos))

	vulneravel := pior < a.limiteRelativo
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevCritica
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"complexidade linear relativa: média=%.1f%%, pior=%.1f%% de %d bits (esperado ~50%%; limite>=%.0f%%)",
			media*100, pior*100, a.bitsPorAmostra, a.limiteRelativo*100,
		),
		Dados: map[string]any{"media": media, "pior": pior},
	}, nil
}

// paraBits converte bytes em bits, do mais significativo ao menos.
func (a *AtaqueComplexidadeLinear) paraBits(bytes []byte, limite int) []int {
	bits := make([]int, 0, limite)
	for i := 0; i < len(bytes) && len(bits) < limite; i++ {
		by := bytes[i]
		for b := 7; b >= 0; b-- {
			bits = append(bits, int((by>>b)&1))
		}
	}
	if len(bits) > limite {
		bits = bits[:limite]
	}
	return bits
}

// berlekampMassey calcula a complexidade linear sobre GF(2).
func (a *AtaqueComplexidadeLinear) berlekampMassey(s []int) int {
	n := len(s)
	c := make([]int, n)
	c[0] = 1
	b := make([]int, n)
	b[0] = 1
	l := 0
	m := -1

	for i := 0; i < n; i++ {
		d := s[i]
		for j := 1; j <= l; j++ {
			d ^= c[j] & s[i-j]
		}

		if d == 1 {
			t := make([]int, n)
			copy(t, c)
			shift := i - m
			for j := 0; j+shift < n; j++ {
				c[j+shift] ^= b[j]
			}
			if l <= i/2 {
				l = i + 1 - l
				m = i
				copy(b, t)
			}
		}
	}

	return l
}
